pkg/scheduler/core/generic_scheduler.go

```diff
@@ -167,52 +167,48 @@ func (g *genericScheduler) selectClusters(clustersScore framework.ClusterScoreLi
 }
 
 func (g *genericScheduler) assignReplicas(
+	ctx context.Context,
 	clusters []*clusterv1alpha1.Cluster,
 	replicaSchedulingStrategy *policyv1alpha1.ReplicaSchedulingStrategy,
 	object *workv1alpha2.ResourceBindingSpec,
 ) ([]workv1alpha2.TargetCluster, error) {
-	startTime := time.Now()
-	defer metrics.ScheduleStep(metrics.ScheduleStepAssignReplicas, startTime)
+	defer metrics.ScheduleStep(metrics.ScheduleStepAssignReplicas, time.Now())
 	if len(clusters) == 0 {
 		return nil, fmt.Errorf("no clusters available to schedule")
 	}
-	targetClusters := make([]workv1alpha2.TargetCluster, len(clusters))
 
 	if object.Replicas > 0 && replicaSchedulingStrategy != nil {
+		var strategy string
+
 		switch replicaSchedulingStrategy.ReplicaSchedulingType {
-		// 1. Duplicated Scheduling
 		case policyv1alpha1.ReplicaSchedulingTypeDuplicated:
-			for i, cluster := range clusters {
-				targetClusters[i] = workv1alpha2.TargetCluster{Name: cluster.Name, Replicas: object.Replicas}
-			}
-			return targetClusters, nil
-		// 2. Divided Scheduling
+			strategy = DuplicatedStrategy
 		case policyv1alpha1.ReplicaSchedulingTypeDivided:
 			switch replicaSchedulingStrategy.ReplicaDivisionPreference {
-			// 2.1 Weighted Scheduling
+			case policyv1alpha1.ReplicaDivisionPreferenceAggregated:
+				strategy = AggregatedStrategy
 			case policyv1alpha1.ReplicaDivisionPreferenceWeighted:
-				// If ReplicaDivisionPreference is set to "Weighted" and WeightPreference is not set,
-				// scheduler will weight all clusters averagely.
-				if replicaSchedulingStrategy.WeightPreference == nil {
-					replicaSchedulingStrategy.WeightPreference = getDefaultWeightPreference(clusters)
+				if replicaSchedulingStrategy.WeightPreference != nil && len(replicaSchedulingStrategy.WeightPreference.DynamicWeight) != 0 {
+					strategy = DynamicWeightStrategy
+				} else {
+					strategy = StaticWeightStrategy
 				}
-				// 2.1.1 Dynamic Weighted Scheduling (by resource)
-				if len(replicaSchedulingStrategy.WeightPreference.DynamicWeight) != 0 {
-					return divideReplicasByDynamicWeight(clusters, replicaSchedulingStrategy.WeightPreference.DynamicWeight, object)
-				}
-				// 2.1.2 Static Weighted Scheduling
-				return divideReplicasByStaticWeight(clusters, replicaSchedulingStrategy.WeightPreference.StaticWeightList, object.Replicas)
-			// 2.2 Aggregated scheduling (by resource)
-			case policyv1alpha1.ReplicaDivisionPreferenceAggregated:
-				return divideReplicasByResource(clusters, object, policyv1alpha1.ReplicaDivisionPreferenceAggregated)
-			default:
-				return nil, fmt.Errorf("undefined replica division preference: %s", replicaSchedulingStrategy.ReplicaDivisionPreference)
 			}
-		default:
-			return nil, fmt.Errorf("undefined replica scheduling type: %s", replicaSchedulingStrategy.ReplicaSchedulingType)
 		}
+
+		assign, ok := assignFuncMap[strategy]
+		if !ok {
+			return nil, fmt.Errorf("unsupported strategy")
```

Comment:
Right, we have the `+kubebuilder:validation` tag in API. So the strategy type should always be determined. But I'm afraid there could be some more types in the future, so return error in case assign function not found. Maybe we can add a more readable message?

Reply:
Maybe we can add a more readable message?

+1